pkg/transactionBackfiller/backfiller.go

```
@@ -296,3 +312,127 @@ func (t *TransactionBackfiller) ProcessBlock(ctx context.Context, blockNumber ui
+			Log:         log,
+			BlockNumber: blockNum,
+			TxHash:      txHash,
+		}
+
+		if _, exists := logsByBlock[blockNum]; !exists {
+			logsByBlock[blockNum] = make([]*LogWithMetadata, 0)
+		}
+		logsByBlock[blockNum] = append(logsByBlock[blockNum], logWithMetadata)
+	}
+
+	// Create worker pool for block fetching and processing
+	wg := &sync.WaitGroup{}
+	blockQueue := make(chan uint64, len(interestingBlocks))
+	errorsRecv := make(chan error, len(interestingBlocks))
+
+	// Launch workers
+	wg.Add(t.config.Workers)
+	for i := 0; i < t.config.Workers; i++ {
+		go func() {
+			defer wg.Done()
+			for blockNum := range blockQueue {
+				logsForBlock, exists := logsByBlock[blockNum]
+				if !exists {
+					continue // No logs for this block, skip
+				}
+
+				// Fetch the block
+				fetchedBlock, err := t.fetcher.FetchBlock(ctx, blockNum)
+				if err != nil {
+					t.logger.Sugar().Errorw("Error fetching block",
+						zap.Uint64("blockNumber", blockNum),
+						zap.Error(err))
+					errorsRecv <- err
+					continue
+				}
+
+				// Process logs for this block
+				t.processLogsForBlock(queueMessage, logsForBlock, fetchedBlock, errorsRecv)
+			}
+		}()
+	}
+
+	// Queue the blocks
+	t.logger.Sugar().Infow("Queueing blocks for processing", zap.Int("count", len(interestingBlocks)))
+	for _, blockNum := range interestingBlocks {
+		blockQueue <- blockNum
+	}
+	close(blockQueue)
+
+	t.logger.Sugar().Infow("Waiting for workers to finish")
+	wg.Wait()
+
+	t.logger.Sugar().Infow("All workers finished, compiling errors")
+	close(errorsRecv)
+	errors := make([]error, 0)
+	for err := range errorsRecv {
+		errors = append(errors, err)
+	}
+
+	if len(errors) > 0 {
+		response.Errors = errors
+		return response
+	}
+
+	return response
+}
+
+// processLogsForBlock processes all logs for a specific block
+func (t *TransactionBackfiller) processLogsForBlock(
+	queueMessage *BackfillerMessage,
+	logs []*LogWithMetadata,
+	fetchedBlock *fetcher.FetchedBlock,
+	errorsChan chan<- error,
```

Rather than dealing with this channel, this function can aggregate a list of errors and simply return it. Alternatively, you'd probably want to return the first error encountered, otherwise you may end up with partial data handled